File: javascript/saddle-points/saddle-points.js
export class Matrix {
    constructor(input) {
        this.rows = input.split('\n').map((line) => line.trim().split(' ').map((i) => Number(i)));
        if (this.rows.length !== 0) {
            this.columns = this.rows[0].map((_, colIdx) => this.rows.map((row) => row[colIdx]));
        } else {
            this.columns = [];
        }

        this.saddlePoints = [];
        for (let rowIdx = 0; rowIdx < this.rows.length; rowIdx++) {
            for (let colIdx = 0; colIdx < this.columns.length; colIdx++) {
                const pointValue = this.rows[rowIdx][colIdx];
                if (this.rows.every((row) => row[colIdx] >= pointValue) && this.columns.every((col) => col[rowIdx] <= pointValue)) {
                    this.saddlePoints.push([rowIdx, colIdx]);
                }
            }
        }
    }
}

File: ruby/saddle-points/saddle_points.rb
class Matrix
  attr_reader :rows, :columns

  def initialize(input)
    @rows = input.lines.map { |line| line.strip.split(" ").map(&:to_i) }
    @columns = @rows.transpose
  end

  def saddle_points
    rows.flat_map.with_index do |row, row_num|
      row.flat_map.with_index do |e, col_num|
        saddle_point = row.all? { |n| n <= e } && columns[col_num].all? { |n| n >= e }
        saddle_point ? [[row_num, col_num]] : []
      end
    end
  end
end
